pandas/formats/format.py

```diff
@@ -1575,13 +1586,22 @@ def _save_chunk(self, start_i, end_i):
                 self.data[col_loc] = col
 
         ix = data_index.to_native_types(slicer=slicer, na_rep=self.na_rep,
+                                        bytes_encoding=self.bytes_encoding,
                                         float_format=self.float_format,
                                         decimal=self.decimal,
                                         date_format=self.date_format,
                                         quoting=self.quoting)
 
         lib.write_csv_rows(self.data, ix, self.nlevels, self.cols, self.writer)
 
+    def _bytes_to_str(self, values):
+        """Modify values list by decoding bytes to str."""
```

Comment:
I tried using `lib.is_bytes_array(arr)` now.  The problem is, it returns False if there are NaNs.  So an array with all bytes apart from some "missing" values is not converted.  This doesn't work, because CSV output by default prints NaN elements in string columns as empty strings.